flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java

```diff
@@ -730,15 +748,49 @@ public CompletableFuture<CoordinationResponse> deliverCoordinationRequestToCoord
                                                         + " does not exist")));
     }
 
+    @Override
+    public JobResourceRequirements requestJobResourceRequirements() {
+        final JobResourceRequirements.Builder builder = JobResourceRequirements.newBuilder();
+        for (JobInformation.VertexInformation vertex : jobInformation.getVertices()) {
+            builder.setParallelismForJobVertex(vertex.getJobVertexID(), 1, vertex.getParallelism());
+        }
+        return builder.build();
+    }
+
+    @Override
+    public void updateJobResourceRequirements(JobResourceRequirements jobResourceRequirements) {
+        if (executionMode == SchedulerExecutionMode.REACTIVE) {
+            throw new UnsupportedOperationException(
+                    "Cannot change the parallelism of a job running in reactive mode.");
+        }
+        final Optional<VertexParallelismStore> maybeUpdateVertexParallelismStore =
+                DefaultVertexParallelismStore.applyJobResourceRequirements(
+                        jobInformation.getVertexParallelismStore(), jobResourceRequirements);
+        if (maybeUpdateVertexParallelismStore.isPresent()) {
+            this.jobInformation =
+                    new JobGraphJobInformation(jobGraph, maybeUpdateVertexParallelismStore.get());
```

Comment:
Maybe we need a new SlotAssigner to assign slots from entire task executor first. Otherwise, the resource(task executor) could not release after job scale down.

This could implement in other issues.

Reply:
IMO, when trigger scales down, it means jobs don't need as many resources.
About speeding up rescale operation. To scale down, throwing away task managers or keeping them will not affect the speed of scale. To scale up, I think it's better to let SlotManager/ResourceManager reserve some resources, so Scheduler and SlotPool just need to declare their resource requirements.

Please correct me if I am wrong.

Reply:
> To scale down, throwing away task managers or keeping them will not affect the speed of scale

Of course it does. Image you have 2 TMs with 2 vertices each (p=4), now you scale down to p=2. If you use 2 TMs with 1 vertex each you can locally recover _all_ state. If you throw away 1 TM, you need to download half the state from the checkpoint directory.

Reply:
Thanks for your explanation. Maybe we need an option to control the behavior of SlotAssigner